src/plugins/cpptemplate/cpptemplate.cpp

```diff
@@ -0,0 +1,111 @@
+/**
+ * @file
+ *
+ * @brief Source for cpptemplate plugin
+ *
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ *
+ */
+
+#include "cpptemplate.hpp"
+#include "cpptemplate_delegate.hpp"
+
+#include <kdbhelper.h>
+
+using elektra::CppTemplateDelegate;
+
+using CppKey = kdb::Key;
+using CppKeySet = kdb::KeySet;
+
+namespace
+{
+
+/**
+ * @brief This function returns a key set containing the plugin contract.
+ *
+ * @return A key set specifying the capabilities of the plugin
+ */
+CppKeySet getContract ()
+{
+	return CppKeySet{ 30,
+			  keyNew ("system/elektra/modules/cpptemplate", KEY_VALUE, "cpptemplate plugin waits for your orders", KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports", KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports/open", KEY_FUNC, elektraCppTemplateOpen, KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports/close", KEY_FUNC, elektraCppTemplateClose, KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports/get", KEY_FUNC, elektraCppTemplateGet, KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports/set", KEY_FUNC, elektraCppTemplateSet, KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports/error", KEY_FUNC, elektraCppTemplateError, KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports/checkconf", KEY_FUNC, elektraCppTemplateCheckConfig, KEY_END),
+#include ELEKTRA_README (cpptemplate)
+			  keyNew ("system/elektra/modules/cpptemplate/infos/version", KEY_VALUE, PLUGINVERSION, KEY_END),
+			  KS_END };
+}
+
+} // end namespace
+
+extern "C" {
+
+typedef Delegator<CppTemplateDelegate> delegator;
+
+/** @see elektraDocOpen */
+int elektraCppTemplateOpen (Plugin * handle, Key * key)
```

Comment:
I think nearly everything of this file below this point could be moved into a macro. No need to duplicate it, the functions are not supposed to be changed by users, aren't they?

And if this file gets much shorter: Is it necessary to split this file with the implementation of the delegator?

Reply:
> I think nearly everything of this file below this point could be moved into a macro. No need to duplicate it, the functions are not supposed to be changed by users, aren't they?

While I agree that the functions could be moved into a macro, I think it also makes sense to show the implementation to the plugin developer. If we use a macro here, then a plugin developer might just assume their is some weird magic involved here. As it currently stands a new plugin developer can look at the template code and he will probably know what is going on.

> And if this file gets much shorter: Is it necessary to split this file with the implementation of the delegator?

Not really. However, this template only includes the basic skeleton of a plugin. The code of an actual plugin might be much longer. For this situation it makes sense to keep the code separated in several files.

Anyway, we can always modify the plugin code after we call `copy-template`. It should be easy for a developer to move the whole code into a single file if she or he prefers it that way.

Reply:
>  I think it also makes sense to show the implementation to the plugin developer. If we use a macro here, then a plugin developer might just assume their is some weird magic involved here.

I cannot follow the argumentation. So we always should prefer to copy code instead of invoking existing code?

> As it currently stands a new plugin developer can look at the template code and he will probably know what is going on.

Why would macros change that?

> Anyway, we can always modify the plugin code after we call copy-template.

Yes, that is why *only* parts that should be changed should be copied around via copy-template.

Reply:
> I cannot follow the argumentation. So we always should prefer to copy code instead of invoking existing code?

Nope. in my opinion macros can make code harder to read and to debug (bad error messages). I also think that plugin developers might want to change the definition of `elektraCppTemplateOpen` or `elektraCppTemplateClose` later. 

Anyway, if you want you can still change the template code after it is merged. You can also close this PR if you do not like the code.

> Why would macros change that?

While it is certainly possible to look up macros, I usually do not do that 😊. I also prefer normal functions over macros since macros cause the compiler to print very bad error messages.

> Yes, that is why *only* parts that should be changed should be copied around via copy-template.

While I agree, I also think that a plugin developer still want to adopt `elektraCppTemplateClose` and `elektraCppTemplateClose` her or his liking. Regardless, **please feel free to close this PR** if you do not like the current code.

Reply:
> in my opinion macros can make code harder to read and to debug (bad error messages).

Macros sometimes caused really bad error messages in earlier versions of GCC. But as these macros are only trivial wrappers, they are unlikely to be important for debugging.

> I also think that plugin developers might want to change the definition of elektraCppTemplateOpen or elektraCppTemplateClose later.

I was hoping that it will be enough for C++ plugins to implement the delegator and that the low-level C functions never need to be touched.

> Regardless, please feel free to close this PR if you do not like the current code.

I like the direction of the PR. I only try to push the code a bit further so that plugins that stem from it get easier to maintain. But also in its current form it will help developers.